File: the-method-of-programming/ch01-string/exercises/10.cpp
#include <iostream>

struct LinkedList {
  int value;
  LinkedList *next;
};

void printLinkedList(const LinkedList *ll) {
  while (ll != 0) {
    std::cout << ll->value << " ";
    ll = ll->next;
  }
  std::cout << std::endl;
}

LinkedList *createLinkedListFromArray(const int *arr, const int n) {
  LinkedList *head = new LinkedList;
  LinkedList *node = head;

  for (int i = 0; i < n; ++i) {
    node->value = arr[i];
    node->next = new LinkedList;
    node = node->next;
  }

  return head;
}

void calcComplementSet(LinkedList* &setA, const LinkedList *setB) {
  LinkedList *head = new LinkedList;
  head->next = setA;
  const LinkedList *b = setB;

  while (b != 0) {
    LinkedList *prev = head;
    LinkedList *a = setA;

    while (a != 0) {
      if (a->value == b->value) {
        prev->next = a->next;
        break;
      } else {
        prev = a;
      }
      a = a->next;
    }

    b = b->next;
  }

  setA = head->next;
}

int main() {
  const int listA[] = { 5, 10, 20, 15, 25, 30 };
  LinkedList *setA = createLinkedListFromArray(listA, 6);
  const int listB[] = { 5, 15, 35, 25 };
  LinkedList *setB = createLinkedListFromArray(listB, 6);
  calcComplementSet(setA, setB);
  printLinkedList(setA);

  return 0;
}

File: the-method-of-programming/ch01-string/exercises/10.js
'use strict'

const assert = require('assert')

class LinkedList {
  constructor(value, next = null) {
    this.value = value
    this.next = next
  }

  static createFromArray(arr) {
    return arr.slice(0, -1).reduceRight(
      (next, elem) => new LinkedList(elem, next),
      new LinkedList(arr[arr.length - 1]),
    )
  }

  toArray() {
    const arr = []
    let node = this

    do {
      arr.push(node.value)
      node = node.next
    } while (node !== null)

    return arr
  }
}

function calcComplementSet(setA, setB) {
  const head = new LinkedList(null, setA)
  let b = setB

  while (b !== null) {
    let prev = head
    let a = setA

    while (a !== null) {
      if (a.value === b.value) {
        prev.next = a.next
        break
      } else {
        prev = a
      }
      a = a.next
    }

    b = b.next
  }

  return head.next
}

function main() {
  const setA = LinkedList.createFromArray([5, 10, 20, 15, 25, 30])
  const setB = LinkedList.createFromArray([5, 15, 35, 25])
  const result = calcComplementSet(setA, setB)
  const expected = LinkedList.createFromArray([10, 20, 30])
  assert.deepStrictEqual(result, expected)
}

main()
